src/python/pants/engine/legacy/parser.py

```diff
@@ -140,27 +141,38 @@ def parse(self, filepath: str, filecontent: bytes):
     # Note that this is incredibly poor sandboxing. There are many ways to get around it.
     # But it's sufficient to tell most users who aren't being actively malicious that they're doing
     # something wrong, and it has a low performance overhead.
-    if self._build_file_imports_behavior != 'allow' and 'import' in python:
+    if "globs" in python or (self._build_file_imports_behavior != 'allow' and 'import' in python):
       io_wrapped_python = StringIO(python)
       for token in tokenize.generate_tokens(io_wrapped_python.readline):
-        if token[1] == 'import':
-          line_being_tokenized = token[4]
-          if self._build_file_imports_behavior == 'warn':
-            logger.warning('{} tried to import - import statements should be avoided ({})'.format(
-              filepath,
-              line_being_tokenized
-            ))
-          elif self._build_file_imports_behavior == 'error':
-            raise ParseError(
-              'import statements have been banned, but tried to import: {}'.format(
-                line_being_tokenized
-              )
+        token_str = token[1]
+        lineno, _ = token[2]
+
+        # We have this deprecation here, rather than in `engine/legacy/structs.py` where the
+        # `sources` field is parsed, so that we can refer to the line number and filename as that
+        # information is not preserved.
+        deprecated_conditional(
+          lambda: token_str in ["globs", "rglobs", "zglobs"],
+          entity_description="Using `globs`, `rglobs`, and `zglobs`",
+          removal_version="1.27.0.dev0",
+          hint_message=f"Using deprecated `{token_str}` in {filepath} at line {lineno}. Instead, "
+                       f"use a list of files and globs, like `sources=['f1.py', '*.java']`. "
+                       f"Specify excludes by putting an `!` at the start of the value, like "
+                       f"`!ignore.py`."
+        )
+
+        if token_str == 'import':
+          if self._build_file_imports_behavior == "allow":
+            continue
+          elif self._build_file_imports_behavior == 'warn':
+            logger.warning(
+              f'Import used in {filepath} at line {lineno}. Import statements should '
+              f'be avoided in BUILD files because TODO... Instead, consider TODO...'
```

Comment:
I have no idea how to explain to a user why they should avoid imports and what to do instead. I understand why we don't want to allow them, but why should a user care? (I think Stu explained this on Slack a week ago but I couldn't find it)

Ideas for what to put?

--

Also, why do we allow imports in the first place? Have we considered deprecating that to always error? Path forward:
1) Deprecate ignoring the error
2) Deprecate it being an option - always error.

Reply:
I don't really remember very well why we discourage imports either (although the thought of imports on BUILD files definitely elicits a shiver down my spine), but I know that these kind of messages like "you can't do X" go down easier if we give them the way we recommend solving that problem (e.g. "you can't use imports, instead we recommend that you write your own target types, here is the documentation").

It's annoying enough being told that you can't do something that you could before, if we can point them in the right direction, it helps a lot :)

Reply:
Reasons imports are bad:
1. The imports people _tend_ to try to use in BUILD files break our caching model. e.g. reading a file, or reading an env var - the engine won't invalidate the BUILD file parse if these change, so they'll get stale results.
2. BUILD file parsing needs to be deterministic, and ideally should be relatively fast; it's hard to accidentally violate either without imports; imports make it much easier to violate these properties.

If people really need imports, they should write their code in a pants plugin, because we (at least try to) over-invalidate when those change.

Reply:
Thank you both!! IMO, we should deprecate this behavior. In a followup, I'll deprecate ignoring imports but will keep the ability to warn for a more incremental migration (we're already deprecating A LOT in 1.25.0).